File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:
    vector<vector<int>> res;
    vector<int> list;
    int sum = 0;
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        backtracking(candidates, target, 0);
        
        return res;
    }

    void backtracking(vector<int>& candidates, int target, int start) {
        if (sum == target) {
            vector<int> copy = list;
            res.push_back(list);
            return;
        }


        for (int i = start; i < candidates.size(); i++) {
            if (sum + candidates[i] > target) {
                break;
            }
            list.push_back(candidates[i]);
            sum += candidates[i];
            backtracking(candidates, target, i);
            sum -= candidates[i];
            list.pop_back();
        }
    }
};

File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    List<List<Integer>> res;
    List<Integer> list;
    int sum;
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        res = new ArrayList<>();
        list = new ArrayList<>();
        sum = 0;
        Arrays.sort(candidates);
        backtracking(candidates, 0, target);

        return res;
    }

    private void backtracking(int[] candidates, int start, int target) {
        if (sum == target) {
            res.add(new ArrayList<Integer>(list));
            return;
        }

        for (int i = start; i < candidates.length; i++) {
            if (sum + candidates[i] > target) {
                return;
            }
            list.add(candidates[i]);
            sum += candidates[i];
            backtracking(candidates, i, target);
            sum -= candidates[i];
            list.remove(list.size() - 1);
        }
    }

}